<!DOCTYPE html>
<html>
  <head>
    <title>Giotto</title>
	<meta charset="utf-8">
	<meta name="viewport" content="initial-scale=1.0, user-scalable=no">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <script src="js/jquery.min.js"></script>
    <script src="js/jquery-ui.min.js"></script>
	<link rel="stylesheet" href="css/jquery-ui.min.css">
	<link rel="stylesheet" href="css/bootstrap.4.1.0.min.css">
	<link rel="stylesheet" href="css/carousel.css">
	<link rel="stylesheet" href="css/prism.css">
	<link rel="stylesheet" href="css/giotto.css">
	<script src="js/bootstrap.4.1.0.min.js"></script>
	<script src="js/carousel.js"></script>
	<script src="js/prism.js"></script>
	<script src="js/giotto.js"></script>
  </head>
  <body data-spy="scroll" data-target="#nex2" data-offset="80">
	<nav class="navbar navbar-expand-lg navbar-light bg-light">
	  <a class="navbar-brand" href="#">Giotto</a>
	  <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
	    <span class="navbar-toggler-icon"></span>
      </button>
      <div class="collapse navbar-collapse" id="navbarSupportedContent">
        <ul class="navbar-nav mr-auto">
          <li class="nav-item active">
            <a class="nav-link" href="giotto.html">Home</a>
          </li>
          <li class="nav-item">
            <a class="nav-link" href="giotto.documentation.html">Documentation</a>
          </li>
          <li class="nav-item">
            <a class="nav-link" href="giotto.install.html">Installation</a>
          </li>
          <li class="nav-item">
            <a class="nav-link" href="giotto.dataset.html">Dataset Examples</a>
          </li>
          <li class="nav-item">
            <a class="nav-link" href="giotto.download.html">Download</a>
          </li>
<!--
          <li class="nav-item dropdown">
            <a class="nav-link dropdown-toggle" href="#" id="navbarDropdown" role="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">Dropdown</a>
            <div class="dropdown-menu" aria-labelledby="navbarDropdown">
              <a class="dropdown-item" href="#">Action</a>
              <a class="dropdown-item" href="#">Another action</a>
              <div class="dropdown-divider"></div>
              <a class="dropdown-item" href="#">Something else here</a>
            </div>
          </li>
          <li class="nav-item">
            <a class="nav-link disabled" href="#" tabindex="-1" aria-disabled="true">Disabled</a>
          </li>
-->
        </ul>
    </nav>
    <div class="container-fluid">
	 <div class="row">
		<nav class="col-md-2 d-none d-md-block bg-light sidebar">
			<div class="sidebar-sticky">
				<ul class="nav flex-column">
                    <li class="nav-item">
                        <a class="nav-link active" href="giotto.dataset.html">
                            Main page <span class="sr-only">(current)</span>
                        </a>
                    </li>
                    <li class="nav-item">
                        <a class="nav-link" href="giotto.osmFISH.html">
                            osmFISH dataset
                        </a>
                    </li>
                    <li class="nav-item">
                        <a class="nav-link" href="giotto.slideseq.html">
                            slideSeq part 1
                        </a>
                    </li>
                    <li class="nav-item">
                        <a class="nav-link" href="giotto.slideseq2.html">
                            slideSeq part 2
                        </a>
                    </li>
                    <li class="nav-item">
                        <a class="nav-link" href="giotto.cycif.html">
                            cyCIF dataset
                        </a>
					</li>
                    <li class="nav-item">
                        <a class="nav-link" href="giotto.seqfish.html">
                            seqFISH+ dataset
                        </a>
					</li>
                    <li class="nav-item">
                        <a class="nav-link" href="giotto.codex.html">
                            CODEX spleen
                        </a>
		    </li>
                    <li class="nav-item">
                        <a class="nav-link" href="giotto.visium.brain.html">
                            Visium brain
                        </a>
		    </li>
                    <li class="nav-item">
                        <a class="nav-link" href="giotto.visium.kidney.html">
                            Visium kidney
                        </a>
		    </li>
                    <li class="nav-item">
                        <a class="nav-link" href="giotto.merfish.html">
                            merFISH hypothalamus
                        </a>
		    </li>
                    <li class="nav-item">
                        <a class="nav-link" href="giotto.starmap.html">
                            STARmap dataset
                        </a>
		    </li>
				</ul>
			</div>
		</nav>

		<main role="main" class="col-md-9 ml-sm-auto col-lg-10 pt-3 px-4">
			<div class="d-flex justify-content-between flex-wrap flex-md-nowrap align-items-center pb-2 mb-3 border-bottom">
				<h1 class="h2">osmFISH Dataset </h1>
			</div>
			<nav id="nex2" class="navbar sticky-top navbar-light bg-light">
<!--  <a class="navbar-brand" href="#">Navbar</a> -->
  <ul class="nav nav-pills">
    <li class="nav-item">
      <a class="nav-link active" href="#s_data_loading">Load</a>
    </li>
    <li class="nav-item">
      <a class="nav-link" href="#s_dimension_red">Dim.Reduce</a>
    </li>
    <li class="nav-item">
      <a class="nav-link" href="#s_clustering">Cluster</a>
    </li>
    <li class="nav-item">
      <a class="nav-link" href="#s_differential">DE</a>
    </li>
    <li class="nav-item">
      <a class="nav-link" href="#s_annotation">Annotate</a>
    </li>
    <li class="nav-item">
      <a class="nav-link" href="#s_spatial_grid">Spatial grid</a>
    </li>
    <li class="nav-item">
      <a class="nav-link" href="#s_spatial_genes">Spatial genes</a>
    </li>
    <li class="nav-item">
      <a class="nav-link" href="#s_hmrf">HMRF</a>
    </li>
    <li class="nav-item">
      <a class="nav-link" href="#s_interaction">Interactions</a>
    </li>
  </ul>
</nav>

<!--<div data-spy="scroll" data-target="#nex2" data-offset="0">-->
<h4 id="s_data_loading">Data loading</h4>
<pre><code class="language-R"># this example works with Giotto v.0.1.2
library(Giotto)
# create instructions
my_python_path = "/Users/rubendries/Bin/anaconda3/envs/py36/bin/pythonw"
results_folder = '/my/results/folder/path/'
instrs = createGiottoInstructions(python_path = my_python_path, show_plot = F, return_plot = T, save_plot = T, save_dir = results_folder, plot_format = 'png', dpi = 300, height = 9, width = 9) </code></pre>


<a href="https://www.nature.com/articles/s41592-018-0175-z">Codeluppi et al</a> created a high quality and very sensitive spatial expression dataset
consisting of 33 genes from 4,839 single cells acquired with osmFISH in
the somatosensory mosue cortex.

<pre><code class="language-R"># select the directory where you have saved the osmFISH data
data_dir = '/path/to/directory/of/osmFISH_data/'
## ss cortex expression DATA ##
osm_exprs = read.table(file = paste0(data_dir,'/','osmFISH_prep_expression.txt'))
## prepare cell locations
osm_locs = read.table(file = paste0(data_dir,'/','osmFISH_prep_cell_coordinates.txt'))
osm_locs = osm_locs[rownames(osm_locs) %in% colnames(osm_exprs),]</code></pre>

Normalize using the standard z-score way:
<pre><code class="language-R">osm_test <- normalizeGiotto(gobject = osm_test)</code></pre>

Alternatively, you can normalize the data using the method proposed in osmFISH paper:
<pre><code class="language-R">raw_expr_matrix = osm_test@raw_exprs
norm_genes = (raw_expr_matrix/rowSums(raw_expr_matrix)) * nrow(raw_expr_matrix)
norm_genes_cells = t((t(norm_genes)/colSums(norm_genes)) * ncol(raw_expr_matrix))
osm_test@custom_expr = norm_genes_cells</code></pre>

Add gene &amp; cell statistics
<pre><code class="language-R">osm_test <- addStatistics(gobject = osm_test)</code></pre>

Save plots according to giotto instructions. This will create subfolder and will provide your own plot name
<pre><code class="language-R">spatPlot2D(gobject = osm_test, cell_color = 'ClusterName', save_param = list(save_folder = '2_Gobject', save_name = 'original_clusters', units = 'in'))
spatPlot2D(gobject = osm_test, cell_color = 'Region', save_param = list(save_folder = '2_Gobject', save_name = 'original_regions', units = 'in'))</code></pre>

<a class="btn btn-primary" data-toggle="collapse" href="#collapseCelltype" role="button" aria-expanded="false" aria-controls="collapseCelltype">osmFISH cell types from paper</a>
<div class="collapse" id="collapseCelltype">
<img src="figures/osmFISH/1_original_clusters.png" class="img-fluid" alt="Responsive image"></img>
</div>
<br><br>
<a class="btn btn-primary" data-toggle="collapse" href="#collapseSpatialregion" role="button" aria-expanded="false" aria-controls="collapseSpatialregion">osmFISH spatial regions from paper</a>
<div class="collapse" id="collapseSpatialregion">
<img src="figures/osmFISH/1_original_clusters.png" class="img-fluid" alt="Responsive image"></img>
</div>

<br><br>

<h4 id="s_dimension_red">Dimension reduction</h4>
Select highly variable genes (HVG). Only 33 genes so use all genes.
Next run PCA on expression values (default), and then run UMAP and TSNE on PCA space (default)
<pre><code class="language-R">osm_test <- runPCA(gobject = osm_test, expression_values = 'custom', scale_unit = F)
signPCA(gobject = osm_test, expression_values = 'custom')
plotPCA_2D(osm_test, save_param = list(save_folder = '3_DimRed', save_name = 'PCA_reduction', units = 'in'))
osm_test <- runUMAP(osm_test, dimensions_to_use = 1:31, expression_values = 'custom', n_threads = 2)
plotUMAP_2D(gobject = osm_test,  save_param = list(save_folder = '3_DimRed', save_name = 'UMAP_reduction', units = 'in'))
osm_test <- runtSNE(osm_test, dimensions_to_use = 1:31, perplexity = 70, check_duplicates = F)
plotTSNE_2D(gobject = osm_test,  save_param = list(save_folder = '3_DimRed', save_name = 'tSNE_reduction', units = 'in'))</code></pre>

<a class="btn btn-primary" data-toggle="collapse" href="#collapsePCA" role="button" aria-expanded="false" aria-controls="collapsePCA">PCA plots</a>
<br>
<div class="collapse" id="collapsePCA">
<img src="./figures/osmFISH/2_PCA_screeplot.png"></img>
<img src="./figures/osmFISH/2_PCA_reduction.png" class="img-fluid" alt="Responsive image"></img>
</div>
<br>
<a class="btn btn-primary" data-toggle="collapse" href="#collapseUMAP" role="button" aria-expanded="false" aria-controls="collapseUMAP">UMAP plot</a>
<br>
<div class="collapse" id="collapseUMAP">
<img src="./figures/osmFISH/2_UMAP_reduction.png" class="img-fluid" alt="Responsive image"></img>
</div>
<br>
<a class="btn btn-primary" data-toggle="collapse" href="#collapseTSNE" role="button" aria-expanded="false" aria-controls="collapseTSNE">tSNE plot</a>
<br>
<div class="collapse" id="collapseTSNE">
<img src="./figures/osmFISH/2_tSNE_reduction.png" class="img-fluid" alt="Responsive image"></img>
</div>
<br>

<br>
<br>
<h4 id="s_clustering">Clustering</h4>
Hierarchical clustering
<pre><code class="language-R">osm_test = doHclust(gobject = osm_test, expression_values = 'custom', k = 34)
plotUMAP_2D(gobject = osm_test, cell_color = 'hclust', point_size = 2.5, show_NN_network = F, edge_alpha = 0.05, save_param = list(save_folder = '4_Cluster', save_name = 'UMAP_hclust', units = 'in'))</code></pre>

<img src="./figures/osmFISH/3_UMAP_hclust.png" class="img-fluid" alt="Responsive image"></img>

<br>
Kmeans clustering
<pre><code class="language-R">osm_test = doKmeans(gobject = osm_test, expression_values = 'custom', centers = 32, nstart = 500)
plotUMAP_2D(gobject = osm_test, cell_color = 'kmeans', point_size = 2.5, show_NN_network = F, edge_alpha = 0.05, save_param =  list(save_folder = '4_Cluster', save_name = 'UMAP_kmeans', units = 'in'))</code></pre>

<img src="./figures/osmFISH/3_UMAP_kmeans.png" class="img-fluid" alt="Responsive image"></img>

<br>
Leiden clustering - first build a SNN network
<pre><code class="language-R">osm_test <- createNearestNetwork(gobject = osm_test, dimensions_to_use = 1:31, k = 15)
osm_test <- doLeidenCluster(gobject = osm_test, resolution = 0.05, n_iterations = 100)
plotUMAP_2D(gobject = osm_test, cell_color = 'leiden_clus', point_size = 2.5, show_NN_network = F, edge_alpha = 0.05, save_param = list(save_folder = '4_Cluster', save_name = 'UMAP_leiden', units = 'in'))
# merge small groups based on similarity
leiden_similarities = getClusterSimilarity(osm_test, expression_values = 'custom', cluster_column = 'leiden_clus')
osm_test = mergeClusters(osm_test, expression_values = 'custom', cluster_column = 'leiden_clus', new_cluster_name = 'leiden_clus_m', max_group_size = 30, force_min_group_size = 20)
plotUMAP_2D(gobject = osm_test, cell_color = 'leiden_clus_m', point_size = 2.5, show_NN_network = F, edge_alpha = 0.05, save_param = list(save_folder = '4_Cluster', save_name = 'UMAP_leiden_merged', units = 'in'))</code></pre>


<img src="./figures/osmFISH/3_UMAP_leiden.png" class="img-fluid" alt="Responsive image"></img>
<img src="./figures/osmFISH/3_UMAP_leiden_merged.png" class="img-fluid" alt="Responsive image"></img>
<br>
Show cluster relationships and plot the cluster dendrogram
<pre><code class="language-R">showClusterHeatmap(gobject = osm_test, expression_values = 'custom', cluster_column = 'leiden_clus_m', save_param = list(save_name = 'heatmap', save_folder = '4_Cluster', units = 'cm'), row_names_gp = grid::gpar(fontsize = 6), column_names_gp = grid::gpar(fontsize = 6))
showClusterDendrogram(osm_test, cluster_column = 'leiden_clus_m', h = 1, rotate = T, save_param = list(save_name = 'dendro', save_folder = '4_Cluster', units = 'cm'))
</code></pre>

<img src="./figures/osmFISH/3_leiden_merged_heatmap.png" class="img-fluid" alt="Responsive image"></img>
<img src="./figures/osmFISH/3_leiden_merged_dendrogram.png" class="img-fluid" alt="Responsive image"></img>

<br>
<br>
<h4 id="s_differential">Differential expression</h4>
Split dendrogram nodes. Can be used to find DEGs at each split of a tree
<pre><code class="language-R">dendsplits = getDendrogramSplits(gobject = osm_test, expression_values = 'custom', cluster_column = 'leiden_clus_m')
split_3_markers = findGiniMarkers(gobject = osm_test, expression_values = 'custom', cluster_column = 'leiden_clus_m', group_1 = unlist(dendsplits[3]$tree_1), group_2 = unlist(dendsplits[3]$tree_2))
## Individual populations ##
markers = findMarkers_one_vs_all(gobject = osm_test, method = 'scran', expression_values = 'custom', cluster_column = 'leiden_clus_m', min_genes = 2, rank_score = 2)
## violinplot
topgenes = markers[, head(.SD, 1), by = 'cluster_ID']$gene_ID
violinPlot(osm_test, genes = unique(topgenes), cluster_column = 'leiden_clus_m', expression_values = 'custom', strip_text = 5, strip_position = 'right', save_param = c(save_name = 'violinplot', save_folder = '6_DEG'))
## cluster heatmap
ranked_genes = c('Bmp4', 'Itpr2', 'Tmem2', 'Ctps', 'Plp1', 'Sox10','Foxj1', 'Aldoc', 'Gfap', 'Acta2', 'Mrc1', 'Vtn', 'Crhbp', 'Slc32a1', 'Gad2', 'Syt6', 'Serpinf1', 'Cpne5', 'Lamp5', 'Hexb', 'Kcnip2', 'Tbr1', 'Ttr', 'Apln', 'Anln', 'Crh', 'Vip', 'Cnr1', 'Pthlh', 'Rorb', 'Flt1', 'Mfge8', 'Pdgfra')
plotMetaDataHeatmap(osm_test, expression_values = 'custom', metadata_cols = c('leiden_clus_m'), custom_gene_order = ranked_genes, save_param = c(save_name = 'metaheatmap', save_folder = '6_DEG'))</code></pre>
<img src="./figures/osmFISH/5_violinplot_leiden_merged.png" class="img-fluid" alt="Responsive image"></img>
<img src="./figures/osmFISH/5_cluster_heatmap_leiden_merged.png" class="img-fluid" alt="Responsive image"></img>
<br>

<h4 id="s_annotation">Cell type annotations</h4>
Create vector with names
<pre><code class="language-R">clusters_SS_cortex = c('OOP', 'OL1', 'OL2', 'OL3', 'OL4',
                       'Ependymal', 'unknown', 'Astro_Gfap', 'vSMC', 'Pericytes',
                       'IN1', 'IN2', 'Pyr1', 'Astro', 'IN3',
                       'IN4', 'Pyr2', 'Miglia1', 'IN5', 'Pyr3',
                       'Choroid', 'Vend1', 'OL5', 'IN6', 'IN7',
                       'IN8', 'IN9', 'Pyr4', 'Pyr5', 'Pyr6',
                       'Vend2', 'Astro_Mfge8', 'OPC')
names(clusters_SS_cortex) = c('m_1', '18', 'm_2', 'm_5', 'm_8',
                              'm_10', 'm_21', '9', 'm_17', 'm_19',
                              'm_11', 'm_14', 'm_6', '30', 'm_3',
                              'm_16', 'm_7', 'm_12', '11', '13',
                              'm_15', 'm_18', '27', 'm_20', '20',
                              '17', '31', '33', '22', 'm_4',
                              'm_13', '8', 'm_9')
osm_test = annotateGiotto(gobject = osm_test, annotation_vector = clusters_SS_cortex, cluster_column = 'leiden_clus_m', name = 'leiden_clus_m_types')
spatDimPlot2D(gobject = osm_test, cell_color = 'leiden_clus_m_types',dim_point_size = 2, spatial_point_size = 2, save_param = c(save_name = 'annotation_leiden_merged_first', save_folder = '7_annotation'))</code></pre>

<img src="./figures/osmFISH/6_annotation_leiden_merged_first.png" class="img-fluid" alt="Responsive image"></img>

<br>
<a class="btn btn-primary" data-toggle="collapse" href="#collapseAnnotateCompare" role="button" aria-expanded="false" aria-controls="collapseAnnotateCompare">Compare clusters with osmFISH paper</a>
<br>
<div class="collapse" id="collapseAnnotateCompare">
<pre><code class="language-R">clusters_det_SS_cortex = c('Olig_COP', 'Olig_NF', 'Olig_MF', 'Olig_mat', 'Olig_mat',
                           'Ependymal', 'unknown', 'Astro_Gfap', 'vSMC', 'Pericytes',
                           'Inh_Crhbp', 'Inh_IC', 'Pyr_L6', 'Periv_Macro', 'Pyr_Cpne5',
                           'unknown', 'Pyr_L2/3', 'Microglia', 'Hippocampus', 'Pyr_L5',
                           'Choroid', 'vEnd', 'unknown', 'Inh_Anln', 'Inh_Crh',
                           'Inh_Vip', 'Inh_Pthlh', 'Pyr_Apln', 'Pyr_Kcnip2', 'Pyr_L4',
                           'vEnd', 'Astro_Mfge8', 'Olig_precursor')
names(clusters_det_SS_cortex) = c('m_1', '18', 'm_2', 'm_5', 'm_8',
                                  'm_10', 'm_21', '9', 'm_17', 'm_19',
                                  'm_11', 'm_14', 'm_6', '30', 'm_3',
                                  'm_16', 'm_7', 'm_12', '11', '13',
                                  'm_15', 'm_18', '27', 'm_20', '20',
                                  '17', '31', '33', '22', 'm_4',
                                  'm_13', '8', 'm_9')
osm_test = annotateGiotto(gobject = osm_test, annotation_vector = clusters_det_SS_cortex, cluster_column = 'leiden_clus_m', name = 'det_cell_types')
spatDimPlot2D(gobject = osm_test, cell_color = 'det_cell_types',dim_point_size = 2, spatial_point_size = 2, save_param = c(save_name = 'annotation_leiden_merged_detailed', save_folder = '7_annotation'))</code></pre>

<img src="./figures/osmFISH/6_annotation_leiden_merged_detailed.png" class="img-fluid" alt="Responsive image"></img>
</div>

<br>
<a class="btn btn-primary" data-toggle="collapse" href="#collapseCoarse" role="button" aria-expanded="false" aria-controls="collapseCoarse">Coarse cell types</a>
<br>
<div class="collapse" id="collapseCoarse">
<pre><code class="language-R">clusters_coarse_SS_cortex = c('Olig', 'Olig', 'Olig', 'Olig', 'Olig',
                              'Ependymal', 'unknown', 'Astro', 'vSMC', 'Pericytes',
                              'Inh', 'Inh', 'Pyr', 'Periv_Macro', 'Pyr',
                              'unknown', 'Pyr', 'Microglia', 'Hippocampus', 'Pyr',
                              'Choroid', 'vEnd', 'unknown', 'Inh', 'Inh',
                              'Inh', 'Inh', 'Pyr', 'Pyr', 'Pyr',
                              'vEnd', 'Astro', 'Olig')
names(clusters_coarse_SS_cortex) = c('Olig_COP', 'Olig_NF', 'Olig_MF', 'Olig_mat', 'Olig_mat',
                                     'Ependymal', 'unknown', 'Astro_Gfap', 'vSMC', 'Pericytes',
                                     'Inh_Crhbp', 'Inh_IC', 'Pyr_L6', 'Periv_Macro', 'Pyr_Cpne5',
                                     'unknown', 'Pyr_L2/3', 'Microglia', 'Hippocampus', 'Pyr_L5',
                                     'Choroid', 'vEnd', 'unknown', 'Inh_Anln', 'Inh_Crh',
                                     'Inh_Vip', 'Inh_Pthlh', 'Pyr_Apln', 'Pyr_Kcnip2', 'Pyr_L4',
                                     'vEnd', 'Astro_Mfge8', 'Olig_precursor')
osm_test = annotateGiotto(gobject = osm_test, annotation_vector = clusters_coarse_SS_cortex, cluster_column = 'det_cell_types', name = 'coarse_cell_types')
spatDimPlot2D(gobject = osm_test, cell_color = 'coarse_cell_types',dim_point_size = 2, spatial_point_size = 2, save_param = c(save_name = 'annotation_leiden_merged_coarse', save_folder = '7_annotation'))</code></pre>
<img src="./figures/osmFISH/6_annotation_leiden_merged_coarse.png" class="img-fluid" alt="Responsive image"></img>
</div>

<h4 id="s_spatial_grid">Spatial grid</h4>
<pre><code class="language-R">osm_test <- createSpatialGrid(gobject = osm_test, sdimx_stepsize = 2000, sdimy_stepsize = 2000, minimum_padding = 0)
spatPlot2D(osm_test, cell_color = 'det_cell_types', show_grid = T, grid_color = 'lightblue', spatial_grid_name = 'spatial_grid', save_param = c(save_name = 'grid_det_cell_types', save_folder = '8_grid'))
</code></pre>
<img src="./figures/osmFISH/7_grid_det_cell_types.png" class="img-fluid" alt="Responsive image"></img>

<br>
Spatial patterns
<pre><code class="language-R">pattern_osm = detectSpatialPatterns(gobject = osm_test, expression_values = 'custom', spatial_grid_name = 'spatial_grid', min_cells_per_grid = 5, scale_unit = T, PC_zscore = 1, show_plot = T)
showPattern2D(osm_test, pattern_osm, dimension = 1, point_size = 4, save_param = c(save_name = 'pattern1_pca', save_folder = '8_grid'))
showPatternGenes(osm_test, pattern_osm, dimension = 1, save_plot = T, save_param = c(save_name = 'pattern1_genes', save_folder = '8_grid', base_height = 3, base_width = 3, dpi = 100))</code></pre>

<img src="./figures/osmFISH/7_pattern1_pca.png" class="img-fluid" alt="Responsive image"></img>
<img src="./figures/osmFISH/7_pattern1_pca_genes.png" class="img-fluid" alt="Responsive image"></img>


<pre><code class="language-R">osm_test <- createSpatialNetwork(gobject = osm_test, k = 5)
spatPlot2D(gobject = osm_test, show_network = T, network_color = 'blue', spatial_network_name = 'spatial_network', point_size = 1, cell_color = 'det_cell_types', save_param = c(save_name = 'spatial_network_k10', save_folder = '9_spatial_network'))</code></pre>

<img src="./figures/osmFISH/8_spatial_network_k5.png" class="img-fluid" alt="Responsive image"></img>

<h4 id="s_spatial_genes">Spatial genes</h4>
There are three ways of doing spatial gene detection (by kmtest, by ranktest, and by silhouette metric)
<pre><code class="language-R">kmtest = binGetSpatialGenes(osm_test, bin_method = 'kmeans', do_fisher_test = T, community_expectation = 5, spatial_network_name = 'spatial_network', verbose = T)
ranktest = binGetSpatialGenes(osm_test, bin_method = 'rank', do_fisher_test = T, community_expectation = 5, spatial_network_name = 'spatial_network', verbose = T)
spatial_genes = calculate_spatial_genes_python(gobject = osm_test, expression_values = 'scaled', python_path = my_python_path, rbp_p=0.99, examine_top=0.1)

spatDimGenePlot2D(osm_test, expression_values = 'normalized', genes = c('Rorb', 'Syt6', 'Gfap', 'Kcnip2'), plot_alignment = 'vertical', cow_n_col = 4, genes_high_color = 'red', genes_mid_color = 'white', genes_low_color = 'darkblue', midpoint = 4, save_param = c(save_name = 'spatial_genes_norm', save_folder = '10_spatial_genes', base_width = 16))</code></pre>


<img src="./figures/osmFISH/9_spatial_network_k5_genes.png" class="img-fluid" alt="Responsive image"></img>

<h4 id="s_hmrf">Hidden Markov random field</h4>
<pre><code class="language-R">my_spatial_genes = spatial_genes[1:20]$genes
my_spatial_genes
 [1] "Rorb"    "Syt6"    "Gfap"    "Lamp5"   "Plp1"    "Sox10"   "Cpne5"
 [8] "Kcnip2"  "Foxj1"   "Tbr1"    "Cnr1"    "Slc32a1" "Hexb"    "Itpr2"
[15] "Ttr"     "Anln"    "Ctps"    "Gad2"    "Mfge8"   "Flt1"</code></pre>

Do HMRF with different betas
<pre><code class="language-R">HMRF_spatial_genes = doHMRF(gobject = osm_test, expression_values = 'normalized', spatial_genes = my_spatial_genes, k = 10, betas = c(0, 0.5, 10), output_folder = paste0(hmrf_folder, '/', 'Spatial_genes/SG_top10_k10_scaled'), python_path = my_python_path, zscore="rowcol", tolerance=1e-5)</code></pre>

View results of HMRF
<pre><code class="language-R">viewHMRFresults(gobject = osm_test, HMRFoutput = HMRF_spatial_genes, k = 10, betas_to_view = seq(0, 5, by = 0.5), point_size = 1)
## add HMRF result of interest to giotto object
osm_test = addHMRF(gobject = osm_test, HMRFoutput = HMRF_spatial_genes, k = 10, betas_to_add = c(0, 0.5, 1.0, 1.5, 2.0), hmrf_name = 'HMRF')</code></pre>

Visualize b = 0, no information from cell neighbors
<pre><code class="language-R">spatPlot2D(gobject = osm_test, cell_color = 'HMRF_k10_b.0', point_size = 1, save_param = c(save_name = 'HMRF_k10_b.0', save_folder = '11_HMRF'))</code></pre>

<img src="./figures/osmFISH/10.osmfish.beta0.png" class="img-fluid" alt="Responsive image"></img>
<br>

Visualize b = 1.0
<pre><code class="language-R">spatPlot2D(gobject = osm_test, cell_color = 'HMRF_k10_b.1.0', point_size = 1, save_param = c(save_name = 'HMRF_k10_b.1.0', save_folder = '11_HMRF'))</code></pre>

<img src="./figures/osmFISH/10.osmfish.beta1.png" class="img-fluid" alt="Responsive image"></img>


<h4 id="s_interaction">Cell cell preferential proximity</h4>
Calculate frequently seen proximities
<pre><code class="language-R">cell_proximities = cellProximityEnrichment(gobject = osm_test, cluster_column = 'det_cell_types', spatial_network_name = 'spatial_network', number_of_simulations = 400)
## barplot
cellProximityBarplot(gobject = osm_test, CPscore = cell_proximities, min_orig_ints = 25, min_sim_ints = 25, save_param = c(save_name = 'barplot_cell_cell_enrichment', save_folder = '12_cell_proxim'))</code></pre>

<img src="./figures/osmFISH/11_barplot_cell_cell_enrichment.png" class="img-fluid" alt="Responsive image"></img>

<br>

<pre><code class="language-R">cellProximityHeatmap(gobject = osm_test, CPscore = cell_proximities, order_cell_types = T, scale = T, color_breaks = c(-1.5, 0, 1.5), color_names = c('blue', 'white', 'red'), save_param = c(save_name = 'heatmap_cell_cell_enrichment', save_folder = '12_cell_proxim', unit = 'in'))</code></pre>

<img src="./figures/osmFISH/11_heatmap_cell_cell_enrichment.png" class="img-fluid" alt="Responsive image"></img>

<br>
<pre><code class="language-R">cellProximityNetwork(gobject = osm_test, CPscore = cell_proximities, remove_self_edges = T, only_show_enrichment_edges = T, save_param = c(save_name = 'network_cell_cell_enrichment', save_folder = '12_cell_proxim'))</code></pre>

<img src="./figures/osmFISH/11_network_cell_cell_enrichment.png" class="img-fluid" alt="Responsive image"></img>

<br>
Visualization
<pre><code class="language-R">spec_interaction = "Astro_Gfap--Olig_mat"
cellProximitySpatPlot2D(gobject = osm_test, interaction_name = spec_interaction, cluster_column = 'det_cell_types', cell_color = 'det_cell_types', coord_fix_ratio = 0.5, point_size_select = 4, point_size_other = 2, save_param = c(save_name = 'cell_cell_enrichment_selected', save_folder = '12_cell_proxim'))</code></pre>

<img src="./figures/osmFISH/11_cell_cell_enrichment_selected.png" class="img-fluid" alt="Responsive image"></img>



		</main>			
 	  </div>
	</div>
  </body>
</html>
