<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: findMarkers_one_vs_all</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for findMarkers_one_vs_all {Giotto}"><tr><td>findMarkers_one_vs_all {Giotto}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>findMarkers_one_vs_all</h2>

<h3>Description</h3>

<p>Identify marker genes for all clusters in a one vs all manner.
</p>


<h3>Usage</h3>

<pre>
findMarkers_one_vs_all(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  cluster_column,
  subset_clusters = NULL,
  method = c("scran", "gini", "mast"),
  pval = 0.01,
  logFC = 0.5,
  min_genes = 10,
  min_expr_gini_score = 0.5,
  min_det_gini_score = 0.5,
  detection_threshold = 0,
  rank_score = 1,
  adjust_columns = NULL,
  verbose = TRUE,
  ...
)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>gobject</code></td>
<td>
<p>giotto object</p>
</td></tr>
<tr valign="top"><td><code>expression_values</code></td>
<td>
<p>gene expression values to use</p>
</td></tr>
<tr valign="top"><td><code>cluster_column</code></td>
<td>
<p>clusters to use</p>
</td></tr>
<tr valign="top"><td><code>subset_clusters</code></td>
<td>
<p>selection of clusters to compare</p>
</td></tr>
<tr valign="top"><td><code>method</code></td>
<td>
<p>method to use to detect differentially expressed genes</p>
</td></tr>
<tr valign="top"><td><code>pval</code></td>
<td>
<p>scran &amp; mast: filter on minimal p-value</p>
</td></tr>
<tr valign="top"><td><code>logFC</code></td>
<td>
<p>scan &amp; mast: filter on logFC</p>
</td></tr>
<tr valign="top"><td><code>min_genes</code></td>
<td>
<p>minimum genes to keep per cluster, overrides pval and logFC</p>
</td></tr>
<tr valign="top"><td><code>min_expr_gini_score</code></td>
<td>
<p>gini: filter on minimum gini coefficient for expression</p>
</td></tr>
<tr valign="top"><td><code>min_det_gini_score</code></td>
<td>
<p>gini: filter minimum gini coefficient for detection</p>
</td></tr>
<tr valign="top"><td><code>detection_threshold</code></td>
<td>
<p>gini: detection threshold for gene expression</p>
</td></tr>
<tr valign="top"><td><code>rank_score</code></td>
<td>
<p>gini: rank scores to include</p>
</td></tr>
<tr valign="top"><td><code>adjust_columns</code></td>
<td>
<p>mast: column in pDataDT to adjust for (e.g. detection rate)</p>
</td></tr>
<tr valign="top"><td><code>verbose</code></td>
<td>
<p>be verbose</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>
<p>additional parameters for the findMarkers function in scran or zlm function in MAST</p>
</td></tr>
</table>


<h3>Details</h3>

<p>Wrapper for all one vs all functions to detect marker genes for clusters.
</p>


<h3>Value</h3>

<p>data.table with marker genes
</p>


<h3>See Also</h3>

<p><code>findScranMarkers_one_vs_all</code>, <code>findGiniMarkers_one_vs_all</code> and <code>findMastMarkers_one_vs_all</code>
</p>


<h3>Examples</h3>

<pre>
    findMarkers_one_vs_all(gobject)
</pre>

<hr /><div style="text-align: center;">[Package <em>Giotto</em> version 0.1.4 ]</div>
</body></html>
